### PR TITLE
Support `value-of<BackedEnum>`

### PR DESCRIPTION
**Pull request by staabm**
as requested in https://github.com/phpstan/phpstan/pull/6841#issuecomment-1070881102

closes https://github.com/phpstan/phpstan/issues/6775

### CODE REVIEW

**Comment by staabm** [comment]
will add more test-cases from the referenced issue now

**Comment by staabm** [comment]
this part of the initial issue

> Additionally PHPStan does not error on types that have this as array key: array<value-of<BackedEnum>, bool>, which means that it silently accepts incorrect code in this case (key-of has the same issue in this case, I guess that should just not be allowed though).
Playground example: [phpstan.org/r/50d372e7-f7bf-49a2-8456-2eb89d97b7bc](https://phpstan.org/r/50d372e7-f7bf-49a2-8456-2eb89d97b7bc)

sounds like a missing rule-check somewhere... I think this can be fixed in a separate PR (which I can work on after this)?

**Comment by staabm** [comment]
(sorry for trial and error.. have no php 8.1 setup yet locally)

build error is unrelated

File: src/PhpDoc/TypeNodeResolver.php [hunk 1]
```diff
@@ -525,6 +525,20 @@ private function resolveGenericTypeNode(GenericTypeNode $typeNode, NameScope $na
 			return new ErrorType();
 		} elseif ($mainTypeName === 'value-of') {
 			if (count($genericTypes) === 1) { // value-of<ValueType>
+				if ($genericTypes[0] instanceof ObjectType) {
```
Comment (ondrejmirtes) [inline src/PhpDoc/TypeNodeResolver.php]: Please do instanceof TypeWithClassName instead

File: src/PhpDoc/TypeNodeResolver.php [hunk 2]
```diff
@@ -525,6 +525,20 @@ private function resolveGenericTypeNode(GenericTypeNode $typeNode, NameScope $na
...
+					$classReflection = $genericTypes[0]->getClassReflection();
```
Comment (ondrejmirtes) [inline src/PhpDoc/TypeNodeResolver.php]: Please check `$this->getReflectionProvider()->hasClass` + `$this->getReflectionProvider()->getClass` like here instead: https://github.com/phpstan/phpstan-src/blob/8ba5f3499f166c2a4e2d80f2f09d929e6f598ee5/src/PhpDoc/TypeNodeResolver.php#L542-L546

File: src/PhpDoc/TypeNodeResolver.php [hunk 3]
```diff
@@ -525,6 +525,20 @@ private function resolveGenericTypeNode(GenericTypeNode $typeNode, NameScope $na
...
+
+					if ($classReflection !== null && $classReflection->isBackedEnum()) {
+						$enumType = $classReflection->getBackedEnumType();
+
+						if ($enumType !== null) {
+							return $enumType;
+						}
+
+						return new ErrorType();
```
Comment (ondrejmirtes) [inline src/PhpDoc/TypeNodeResolver.php]: It's a problem to return error type here. We might still have an object that is iterable so `value-of` should work for it too. Just remove this line.

File: tests/PHPStan/Rules/PhpDoc/data/value-of-enum.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+{
+	case NL = 'The Netherlands';
+	case US = 'United States';
+}
+
+class Foo {
+	/**
+	 * @param value-of<Country> $countryName
+	 */
+	function hello(string $countryName): void
```
Comment (ondrejmirtes) [inline tests/PHPStan/Rules/PhpDoc/data/value-of-enum.php:23]: Please also add a test that fails, e.g. string typehint + int-backed enum or similar.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Please also add a case to NodeScopeResolverTest that tests what `value-of` resolves to inside a method body.

File: tests/PHPStan/Rules/PhpDoc/data/value-of-enum.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	 * @param value-of<CountryNo> $shouldError
+	 */
+	function helloError2(string $shouldError): void
+	{
+		// ...
+	}
+
+	function doFoo() {
+		$this->hello(Country::NL);
+		$this->hello(CountryNo::NL);
```
Comment (staabm) [inline tests/PHPStan/Rules/PhpDoc/data/value-of-enum.php]: this line does not yet error.. but I think this needs a new Rule or enhancing an existing rule..?

should I keep this line or delete it for now?

-----

another case which should error and does not yet is
```php
	/**
	 * @param array<value-of<Country>, bool> $countryMap
	 */
	function helloArray(array $countryMap): void {
		// ...
	}

	function doFooArray() {
		// should not be allowed, 'abc' does not match value-of<Country>
		$this->helloArray(['abc' => true]);
		$this->helloArray(['abc' => 123]);

		// wrong key type
		$this->helloArray([true]);
	}
```

same question here.. I think this needs a new Rule or enhancing an existing rule..?

Comment (ondrejmirtes) [reply tests/PHPStan/Rules/PhpDoc/data/value-of-enum.php]: I don't understand. You have a test-case in a PHPDoc incompatibility rule. It's not going to report wrong method calls. For that you need to add a test-case to CallMethodsRuleTest.php.

**Comment by ondrejmirtes** [comment]
About this:

> Additionally PHPStan does not error on types that have this as array key: array<value-of, bool>, which means that it silently accepts incorrect code in this case 

Yeah, let's discuss this separately in another PR. The issue here is that `value-of<something wrong>` produces an ErrorType in TypeNodeResolver, and `array<*ERROR*, X>` gets cast to `array<int|string, X>` in TypeNodeResolver but it needs to stay `array<*ERROR*, X>` on this line: https://github.com/phpstan/phpstan-src/blob/14cb258f10aaea2be352b650f40ce36fd7a0d131/src/PhpDoc/TypeNodeResolver.php#L460

File: tests/PHPStan/Rules/PhpDoc/data/value-of-enum.php (hunk 5 above)
Comment (staabm) [reply tests/PHPStan/Rules/PhpDoc/data/value-of-enum.php]: good point. added test-cases in CallMethodsRuleTest

**Comment by ondrejmirtes** [comment]
Please add NodeScopeResolverTest case to verify it works as expected.

**Comment by staabm** [comment]
> Please add NodeScopeResolverTest case to verify it works as expected.

done

File: tests/PHPStan/Analyser/NodeScopeResolverTest.php [hunk 6]
```diff
@@ -800,6 +800,11 @@ public function dataFileAsserts(): iterable
 			yield from $this->gatherAssertTypes(__DIR__ . '/data/bug-6859.php');
 		}
 
+		if (PHP_VERSION_ID >= 80100) {
+			require_once __DIR__ . '/data/value-of-enum.php';
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/NodeScopeResolverTest.php]: Require is not needed here, make sure to run `composer dump` locally first.

File: tests/PHPStan/Analyser/data/value-of-enum.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+	{
+		return Country::US;
+	}
+
+	/**
+	 * @param value-of<Country> $countryName
+	 */
+	function hello($countryName)
+	{
+		assertType('string', $countryName);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/value-of-enum.php]: This isn't good enough, right? Should be `'The Netherlands'|'United States'` really.

Comment (ondrejmirtes) [reply tests/PHPStan/Analyser/data/value-of-enum.php]: It's also why that bug isn't report in CallMethodsRuleTest.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Obviously the implementation in TypeNodeResolver needs to change. Please iterate over `ClassReflection::getEnumCases()` to get to actual scalar values.

File: tests/PHPStan/Analyser/NodeScopeResolverTest.php (hunk 6 above)
Comment (staabm) [reply tests/PHPStan/Analyser/NodeScopeResolverTest.php]: I checked it locally.. the tests do not work without this require line.

even if I run `composer dump` beforehand

File: tests/PHPStan/Analyser/data/value-of-enum.php (hunk 7 above)
Comment (staabm) [reply tests/PHPStan/Analyser/data/value-of-enum.php]: fixed it. good catch.

File: tests/PHPStan/Analyser/NodeScopeResolverTest.php (hunk 6 above)
Comment (staabm) [reply tests/PHPStan/Analyser/NodeScopeResolverTest.php]: ahh.. I need to `/my/php8.1 composer dump` and then it works

File: tests/PHPStan/Rules/Methods/data/call-method-in-enum.php [hunk 8]
```diff
@@ -30,3 +30,43 @@ public function doFoo()
...
+	 * @param array<value-of<Country>, bool> $countryMap
+	 */
+	function helloArray(array $countryMap): void {
+		// ...
+	}
+
+	function doFooArray() {
+		$this->hello(CountryNo::NL);
+
+		// 'abc' does not match value-of<Country>
```
Comment (staabm) [inline tests/PHPStan/Rules/Methods/data/call-method-in-enum.php:65]: as expected, we now see errors here.

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**